### PR TITLE
drivers/at86rf2xx: improve precondition checks on state transition

### PR DESCRIPTION
**Pull request by thomaseichinger**
The rational behind this change is the following:
If the transceiver is in any *_BUSY state when `at86rf2xx_set_state()`
gets called this would bypass the `(state == old_state)` check and
unneeded state transitions could be triggered.

@alexaring could be interested in this too.

### CODE REVIEW

**Comment by daniel-k** [comment]
Good catch! Looks fine.

**Comment by alexaring** [comment]
Hi,

I think the followed function doesn't work for RX_AACK_ON and I don't see that this function isn't called anymore or different for RX_AACK_ON case:

static inline void _set_state(at86rf2xx_t *dev, uint8_t state)
{
    at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_STATE, state);
    while (at86rf2xx_get_status(dev) != state);
    dev->state = state;
}

The issue is a special case if SFD is detected after moving into RX_AACK_ON:

MCU          -> at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_STATE, state);
at86rf2xx   -> going into RX_AACK_ON
/* here begins the race */
at86rf2xx   -> SFD detected -> going into BUSY of RX_AACK
MCU          -> while (at86rf2xx_get_status(dev) != state);
                  \-> the MCU will stuck inside this loop until BUSY is finisched because it's not RX_AACK_ON,  this is only the case if SFD is detected directly after RX_AACK_ON and the MCU was too slow to get the successful RX_AACK_ON state change.

Don't know if you really fix this handling with the patch, I don't see it. So far _set_state is still called for RX_AACK_ON.

- Alex

**Comment by daniel-k** [comment]
> Don't know if you really fix this handling with the patch, I don't see it. So far _set_state is still called for RX_AACK_ON.

Although I think that you are right, that the described race condition is not fixed by this PR, it wasn't supposed to be fixed here IMO. 

I would be interested if this race can reproduced in hardware, since the MCU is actively polling at several MHz. Do you have observed such behaviour?

**Comment by alexaring** [comment]
On RIOT I didn't saw such race, on Linux and "atusb firmware" yes.
But I also never tried to debug such behaviour on RIOT boards.

In Linux we don't poll to check "if we are into the state which we like to change in", we have an assert there and if it's not match then we print a warning and this warning should never occur. That's how I detected such race.

About polling this value:
For example: at86rf233 page 41 (on paper), It takes ~160 µs to go into this state.
http://www.atmel.com/images/atmel-8351-mcu_wireless-at86rf233_datasheet.pdf
(RX state line)

If RIOT can hold this real time requirement at worst-case scenario at each board configuration, etc. then this race can't happen. I agree :-)

Let me calculate something (Don't trust the calculation here..., I never did something like that):
Assume: SPI clock ~4 Mhz = 4us
So we can get/set 1 Bit in 4 us, we do a set_state and then a get_state, each command takes 16 bits for spi bus.

32 * 4 = 128 us. It would be fit, but then you could interrupted by another hardirq (don't know if irq's are disabled), but it's a small space there.

**Comment by thomaseichinger** [comment]
> Let me calculate something (Don't trust the calculation here..., I never did something like that):
> Assume: SPI clock ~4 Mhz = 4us
> So we can get/set 1 Bit in 4 us, we do a set_state and then a get_state, each command takes 16 bits 
> for spi bus.
> 
> 32 * 4 = 128 us. It would be fit, but then you could interrupted by another hardirq (don't know if irq's are 
> disabled), but it's a small space there.

I think we have to start counting after the first 16 bits (initiating state transition) are transported which would result in 64us. Additionally, taken a transmission to this receiver is started at exactly this time, SHR (preamble + SFD = 10 symbols) transmission takes 160 us. [at86rf231 spec 8.1.1.1] .
This would result in 224us.

But besides of that, can we come up with a solution that is not dependant on the mercy of timing?
For example, is it sufficient for our use case, to report a successful states transition, to poll on ONGOING_STATE_TRANSITION?

**Comment by alexaring** [comment]
Yes, 64 us.

With "ONGOING_STATE_TRANSITION" you mean the "STATE_TRANSITION_IN_PROGRESS"?
I never got such state status when I develop the at86rf230 under linux and the highest clock rate was 6Mhz.

Why not simple handle the one case to ignore RX_AACK_BUSY when going into RX_AACK_ON? That's how we solved it inside Linux kernel. :-)

**Comment by authmillenon** [comment]
Please resolve your discussion until tonight, otherwise the PR will be postponed to the next release.

**Comment by haukepetersen** [comment]
postponed

**Comment by kYc0o** [comment]
Can this be retaken for tomorrow's Hack'n'ACK?

**Comment by kYc0o** [comment]
Do we have news on this one?

**Comment by kYc0o** [comment]
And this will be postponed again...

**Comment by miri64** [comment]
Ping? I'll leave this open, since it is a bugfix and shouldn't be that hard to review...

**Comment by alexaring** [comment]
I don't know if it's really clear what I meant above, I simple show you the workaround which we have in Linux code:

http://lxr.free-electrons.com/source/drivers/net/ieee802154/at86rf230.c#L423

There we simple don't assert to state RX_AACK_ON and we hit the BUSY state, because it can be that the transceiver detects a SFD. On RIOT this problems (if it appears there) ends in the polling of "if we are in the state which we want to change in" if the "race" above occurs, the loop will end until the RX_BUSY state changed to RX_AACK_ON again, but the state change was "maybe" successful before. It's just that the driver doesn't catch the state change and the transceiver changed it on his own.

-Alex

**Comment by miri64** [comment]
I'm still not sure I understand the answer ^^ (just because I don't know much about the state transitions of this device in detail).

Two things I catch from this discussion:

@daniel-k said

> Although I think that you are right, that the described race condition is not fixed by this PR, it wasn't supposed to be fixed here IMO.

But @alexaring said:

> Why not simple handle the one case to ignore RX_AACK_BUSY when going into RX_AACK_ON? That's how we solved it inside Linux kernel. :-)

@thomaseichinger does this make sense to you in the context of this PR? If it does, please address it, if not I would say let's merge this PR as is.

**Comment by miri64** [comment]
(also, considering the age of this PR a rebase might not be a bad thing ;-))

**Comment by thomaseichinger** [comment]
@miri64 Updated and rebased.

**Comment by miri64** [comment]
@alexaring stated in the IRC that he is alright with the current changes.

**Comment by miri64** [comment]
Will test tomorrow.

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 1]
```diff
@@ -421,17 +421,23 @@ void at86rf2xx_set_option(at86rf2xx_t *dev, uint16_t option, bool state)
 static inline void _set_state(at86rf2xx_t *dev, uint8_t state)
 {
     at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_STATE, state);
-    while (at86rf2xx_get_status(dev) != state);
+
```
Comment (miri64) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: Please remove whitespaces.

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 2]
```diff
@@ -421,17 +421,23 @@ void at86rf2xx_set_option(at86rf2xx_t *dev, uint16_t option, bool state)
 static inline void _set_state(at86rf2xx_t *dev, uint8_t state)
 {
     at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_STATE, state);
-    while (at86rf2xx_get_status(dev) != state);
+    
+    /* To prevent a possible race condition when changing to
+     * RX_AACK_ON state the state doesn't get read back in that
+     * case. See discussion
```
Comment (miri64) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: Please remove trailing whitespaces.

**Comment by miri64** [comment]
Sorry I forgot about this, but I tested it now on IoT-Lab and samr21-xpro, so ACK if requested changes were made.

**Comment by thomaseichinger** [comment]
@miri64 removed the white spaces and immediately squashed.

**Comment by miri64** [comment]
arduino-zero has problems with the arduino-hello_world example (summon @aabadie). I guess this is unrelated, but just for good measure I try it again.

**Comment by miri64** [comment]
Please provide a backport to the 2016.10 release branch.

**Comment by thomaseichinger** [comment]
@miri64 Please see #6062.

**Comment by biboc** [comment]
@thomaseichinger, @alexaring 
Did you try to make the transceiver sleep and wake up again? 
Using NETOPT_STATE_SLEEP(AT86RF2XX_STATE_SLEEP(0x0f)) then NETOPT_STATE_IDLE(AT86RF2XX_STATE_RX_AACK_ON(0x16)), the transceiver does not wake up properly (can't ping from another board)

**Comment by biboc** [comment]
I'm on IRC riot-os to discuss if needed

**Comment by biboc** [comment]
Ok found the problem, will make a PR
DELAY after restart was not big enough

**Comment by gebart** [comment]
@biboc I aber used sleep mode with this driver on a samr21-xpro. I don't remember exactly which PRs and patches were applied though, you could take a look at this branch:
https://github.com/gebart/RIOT/commits/wip/kinetis-pm
There's a ton of unrelated changes there, but some are touching the at86rf2xx driver code.

**Comment by biboc** [comment]
Did a PR #7865 to solve the problem

**Comment by roberthartung** [comment]
*ping* I am trying to implement basic mode. Is the race condition described here also a problem for basic mode?

**Comment by alexaring** [comment]
The race condition always exists when switching to RX_ON/RX_AACK_ON . Why do you want to implement basic operation? You don't have interframe spacing time there (I know that from atmel support - datasheet says nothing about that). For low-level/do everything you maybe want AACK/ARET mode in promiscuous mode.

**Comment by roberthartung** [comment]
For low-power applications I don't want to use any AACK / ARET stuff. Is that possible with extended mode as well (e.g. disabling all these options) - maybe after #8216 is merged, this works now as expected ;)

**Comment by alexaring** [comment]
You can change the CSMA paramters that it does nothing, also disable AACK handling completely (some special frame retries value). AACK can be disable, but you running out of 802.15.4 spec, if you don't handle it correctly over software (if possible).

**Comment by alexaring** [comment]
Put it into promiscuous mode according the datasheet and disable CSMA/CAA over right parameters.

**Comment by roberthartung** [comment]
Well, AACK might be useful still, but ARET not, I will try it tomorrow then and see if that works correctly and what the power consumption in extended mode vs. basic mode is and see that's the difference there.

**Comment by alexaring** [comment]
Did you saw:

11.10 Reduced Power Consumption Mode (RPC) ?

In case of rf233

**Comment by roberthartung** [comment]
Yes, I saw that already. But still Extended Mode has to come at a drawback, why having two modes if there's no drawback?

**Comment by alexaring** [comment]
7.2.4 TX_ARET_ON – Transmit with Automatic Frame Retransmission and CSMA-CA Retry

Note 1: If MAX_CSMA_RETRIES = 7 no retry is
 performed

Also never set the ack request bit.

follow the flow diagramm = for me it looks like basic transmit.

For RX:

AACK_DIS_ACK (seems to be valid only for promiscuous mode, but this will disable address filtering)

**Comment by roberthartung** [comment]
@alexaring Thanks, I already know about the datasheet and all the functions. I've also written my own driver for the basic mode back in the days. I will give it a try and do some measurements.

**Comment by roberthartung** [comment]
> For RX:
> 
> AACK_DIS_ACK (seems to be valid only for promiscuous mode, but this will disable address filtering)

Yes. I also need packets with invalid CRC sum, which for now, I only receive in basic mode.

**Comment by alexaring** [comment]
Or promiscuous, which will for sure disable the address filter which is not what you want for power saving, but then why care about invalid CRC?

**Comment by roberthartung** [comment]
> but then why care about invalid CRC?

Right now for a evaluation I am using RIOT for.

**Pull request merged**